api/src/middleware/rate-limiter-global.ts

```diff
@@ -0,0 +1,53 @@
+import { RequestHandler } from 'express';
+import ms from 'ms';
+import { RateLimiterMemcache, RateLimiterMemory, RateLimiterRedis } from 'rate-limiter-flexible';
+import env from '../env';
+import { HitRateLimitException } from '../exceptions/index';
+import logger from '../logger';
+import { createRateLimiter } from '../rate-limiter';
+import asyncHandler from '../utils/async-handler';
+import { validateEnv } from '../utils/validate-env';
+
+const RATE_LIMITER_GLOBAL_KEY = 'global-rate-limit';
+
+let checkRateLimit: RequestHandler = (_req, _res, next) => next();
+export let rateLimiterGlobal: RateLimiterRedis | RateLimiterMemcache | RateLimiterMemory;
+
+if (env.RATE_LIMITER_GLOBAL_ENABLED === true) {
+	validateEnv(['RATE_LIMITER_GLOBAL_STORE', 'RATE_LIMITER_GLOBAL_DURATION', 'RATE_LIMITER_GLOBAL_POINTS']);
+	validateConfiguration();
+
+	rateLimiterGlobal = createRateLimiter('RATE_LIMITER_GLOBAL');
+
+	checkRateLimit = asyncHandler(async (_req, res, next) => {
+		try {
+			await rateLimiterGlobal.consume(RATE_LIMITER_GLOBAL_KEY, 1);
+		} catch (rateLimiterRes: any) {
+			if (rateLimiterRes instanceof Error) throw rateLimiterRes;
+
+			res.set('Retry-After', String(rateLimiterRes.msBeforeNext / 1000));
```

Comment:
Note: We can merge this PR first and do this separately if that's preferred 👍

This is technically a carry-over from rate limiter IP original code so a slight tangent from the purpose of this PR, but I believe `rateLimiterRes.msBeforeNext / 1000` should actually be rounded/floored for `Retry-After` header's spec which says it should be an integer, similar to the discussion https://github.com/directus/directus/pull/17464#discussion_r1101406835 for `max-age`.

With a local test of point as 1 and duration as 10, I was able to receive decimal places in `Retry-After` response header:

![](https://user-images.githubusercontent.com/42867097/221185979-7366b203-dadc-46ee-8215-7b26adec21e9.png)

Additional References:

- MDN: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After#directives
- Similar "Retry-After header should not have decimal place" issue in dotnet: https://github.com/Microsoft/dotnet/issues/955#issuecomment-476234572

Reply:
I'd say let's make this adjustment now 👍

Reply:
Resolved in https://github.com/directus/directus/pull/17296/commits/3771f28eae15f25ab76bed72c77181529365fa12